File: js/src/get-kth-from-end.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */

// 解法一：快慢指针
var getKthFromEnd1 = function(head, k) {
    let f = head;
    let s = head;

    while (k > 0) {
        f = head.next;
        k--;
    }

    while(f) {
        f = f.next;
        s = s.next;
    }

    return s;
};

// 解法二，转化为 map，取数
var getKthFromEnd2 = function(head, k) {
    const map = new Map();
    let i = 1;


    while (head) {
        map.set(i, head);

        head = head.next;
        i++;
    }

    return map.get(map.size - k + 1);
};

File: rs/src/get_kth_from_end.rs

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
  pub val: i32,
  pub next: Option<Box<ListNode>>
}

#[allow(dead_code)]
pub fn get_kth_from_end_1(head: Option<Box<ListNode>>, k: i32) -> Option<Box<ListNode>> {
    let mut f = head.as_ref();
    let mut s = head.as_ref();

    for _ in 0..k {
        f = f.unwrap().next.as_ref()
    }

    while f.is_some() {
        s = s.unwrap().next.as_ref();
        f = f.unwrap().next.as_ref();
    }

    return s.cloned();
}
